rust/src/hazmat/hkdf.rs

```diff
@@ -0,0 +1,140 @@
+                let mut prk = [0u8; $tag_len];
+                unsafe {
+                    hacl_star_sys::$extract(
+                        prk.as_mut_ptr(),
+                        salt.as_ptr() as _,
+                        salt.len().try_into().unwrap(),
+                        ikm.as_ptr() as _,
+                        ikm.len().try_into().unwrap(),
+                    );
+                }
+                prk
+            }
+
+            /// HKDF expand using the pre-key material `prk` and `info`. The output length
+            /// is defined through the result type.
+            /// Returns the key material in an array of length `okm_len` or
+            /// [`Error::OkmTooLarge`] of the requested `okm_len` is too large.
+            ///
+            /// Note that this function panics if `salt`, `ikm`, or `OKM_LEN` is larger than 2**32 bytes.
+            pub fn expand<const OKM_LEN: usize>(
+                prk: &[u8],
+                info: &[u8],
+            ) -> Result<[u8; OKM_LEN], Error> {
+                if OKM_LEN > 255 * $tag_len {
+                    // Output size is too large. HACL doesn't catch this.
+                    return Err(Error::OkmTooLarge);
+                }
+                let mut okm = [0u8; OKM_LEN];
+                unsafe {
+                    hacl_star_sys::$expand(
+                        okm.as_mut_ptr(),
+                        prk.as_ptr() as _,
+                        prk.len().try_into().unwrap(),
+                        info.as_ptr() as _,
+                        info.len().try_into().unwrap(),
+                        OKM_LEN.try_into().unwrap(),
+                    );
+                }
+                Ok(okm)
+            }
+
+            /// HKDF using the `salt`, input key material `ikm`, `info`. The output length
+            /// is defined through the result type.
+            /// Calls `extract` and `expand` with the given input.
+            ///
+            /// Returns the key material in an array of length `okm_len`.
+            pub fn hkdf<const OKM_LEN: usize>(
+                salt: &[u8],
+                ikm: &[u8],
+                info: &[u8],
+            ) -> Result<[u8; OKM_LEN], Error> {
+                let prk = extract(salt, ikm);
+                expand(&prk, info)
+            }
+
+            /// This module uses heap allocated vectors for cases where the output
+            /// length is not const.
+            pub mod vec {
+                use super::super::Error;
+
+                /// HKDF expand using the pre-key material `prk` and `info`. The output length
+                /// is defined through the result type.
+                /// Returns the key material in an array of length `okm_len` or
+                /// [`Error::OkmTooLarge`] of the requested `okm_len` is too large.
+                ///
+                /// Note that this function panics if `salt`, `ikm`, or `OKM_LEN` is larger than 2**32 bytes.
```

Comment:
Question: As we already return a `Result`: Couldn't we already catch the panic-condition?

Reply:
Good point... I added the `Result` later and forgot about the panic.